### PR TITLE
Resolving Swagger Violations for RecoveryServices

### PR DESCRIPTION
**Pull request by sonathan**
This PR includes:
1. Handling Swagger Violations for RecoveryServices:
    - Property must follow camel casing.
    - x-ms-examples should be added.
    - Patch Should not have any property as required in Request body.
    -Making ListBySubscriptionId and ListByResourceGroup as paginated.
    - id, name should be readonly.

This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process. 

### PR information
- [x] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [x] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [x] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [x] My spec meets the review criteria:
  - [x] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [x] The spec follows the guidelines described in the [Swagger checklist](../documentation/swagger-checklist.md) document.
  - [x] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

### CODE REVIEW

**Comment by sonathan** [comment]
@marstr Could you please review this PR?

**Comment by sonathan** [comment]
@marstr Could you please review this PR?

**Comment by marstr** [comment]
Howdy, sorry for the delay. The holidays interrupted normal operations in the repository. We're here now, I'll take a look.

**Comment by marstr** [comment]
The changes above were rebased to accomodate for the December 2017 refactor of this Repo.

**Comment by marstr** [comment]
I've inspected the 4 linter errors that remain, and can confirm that they are not related the commits in this PR so far.

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (marstr) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: It is unclear to me why `PatchTrackedResource` and `PatchVault` were added, they seem to be identical to their non-patch counterparts.  Can you explain why it is necessary or desirable to have both?

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (marstr) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: Doing this will cause breaking changes. As I mention below, whether or not the new type is necessary is unclear to me.

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/examples/DeleteRegisteredIdentities.json
Comment (marstr) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/examples/DeleteRegisteredIdentities.json]: Is this a real subscription ID? If it is, we should scrub it.

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (marstr) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: This is awkward or wrong grammatically. Maybe, "Link to the next chunk of the response" ?

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/registeredidentities.json
Comment (marstr) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/registeredidentities.json]: Before I approve, I'm going to need to check to ensure that this doesn't cause breaking changes in any of the known generated languages.

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/registeredidentities.json
Comment (marstr) [inline specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/registeredidentities.json]: None of these properties seem to have descriptions, which is something the linter is supposed to be looking for. Is there a reason descriptions are not included?

**Review by marstr — CHANGES_REQUESTED** [review]

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (sonathan) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: Hi @marstr ,

PatchTrackedResource doesn't have "location" as required property as opposed to its counter part TrackedResource. This was the reason(All the top level properties should be optional) we had to introduce a new model for PATCH call: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}" . This was a swagger violation and hence we updated it. This solution was suggested by @amarzavery 


Here is the exact violation you can refer to:


ERROR   (PatchBodyParametersSchema/R2016/RPCViolation): Properties of a PATCH request   body must not be required. PATCH operation: 'Vaults_Update' Model Definition:   'TrackedResource' Property: 'location'
--
- file:///E:/src/azure-rest-api-specs/specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/2016-06-01/vaults.json:569:6   ($.definitions.TrackedResource.required)

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (sonathan) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: Hi @marstr ,

PatchTrackedResource doesn't have "location" as required property as opposed to its counter part TrackedResource. This was the reason(All the top level properties should be optional) we had to introduce a new model for PATCH call: /subscriptions/{subscriptionId}/resourceGroups/{resourceGroupName}/providers/Microsoft.RecoveryServices/vaults/{vaultName}" . This was a swagger violation and hence we updated it. This solution was suggested by @amarzavery

Here is the exact violation you can refer to:

ERROR (PatchBodyParametersSchema/R2016/RPCViolation): Properties of a PATCH request body must not be required. PATCH operation: 'Vaults_Update' Model Definition: 'TrackedResource' Property: 'location'


Also we already have handling for this in our Service code. Hence this won't break anything.

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/examples/DeleteRegisteredIdentities.json
Comment (sonathan) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/examples/DeleteRegisteredIdentities.json]: Yes, This is a real subscriptionId, Updating it. Thanks for your suggestion.

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/registeredidentities.json
Comment (sonathan) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/registeredidentities.json]: Thank-you for letting me know. We should definitely have descriptions in this case.

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (sonathan) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: Okay. :) Thanks

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (marstr) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: Thanks for this explanation, it is exactly what I was looking for.

File: specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json
Comment (marstr) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: > @sonathan says:
Also we have handling for this in our Service code. Hence this won't break anything.

Sorry, I should have been more clear. Changing the type will cause a breaking change for customers consuming some of the generated SDKs, static languages in particular.

Comment (marstr) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: I just had a long chat about this offline with @amarzavery. We agree that our hands are tied here. While your change will indeed cause breaking changes for some of the client languages, the consequences of not making this change are not any more palatable. I've labeled this PR as containing potential SDK breaks, but I won't have this issue prevent my approval.

Comment (sonathan) [reply specification/recoveryservices/resource-manager/Microsoft.RecoveryServices/stable/2016-06-01/vaults.json]: @marstr Thank-You, Is there any change with respect to this PR required from my end?

**Comment by sonathan** [comment]
Hi @marstr 

Is there anything pending from my end with respect to this PR?

File: specification/recoveryservices/resource-manager/readme.md
Comment (marstr) [inline specification/recoveryservices/resource-manager/readme.md]: Looks like this file was the only unique thing about the `2016-12` package, so it makes sense that if the file gets deleted, this whole tag does.

Is there a plan to reintroduce this swagger? Why was it ever included?

**Comment by marstr** [comment]
Thanks for pinging me, @sonathan. I've added a question about why `backup.json` is being deleted. Other than that, this PR looks good to me.

File: specification/recoveryservices/resource-manager/readme.md
Comment (marstr) [reply specification/recoveryservices/resource-manager/readme.md]: Talking with a couple of my colleagues offline, we agree that unless there is something confidential about the contents of this file it should stay included.

Comment (sonathan) [reply specification/recoveryservices/resource-manager/readme.md]: Hi @marstr ,

Thanks for the update. The APIs present in Backup.json is now moved to RecoveryServicesBackup spec file, hence was not needed here.

Comment (marstr) [reply specification/recoveryservices/resource-manager/readme.md]: I see, makes sense.

**Comment by marstr** [comment]
For folks curious about `backup.json` being deleted in this PR, it is actually just a duplicate. The source of truth on this can be found here: https://github.com/Azure/azure-rest-api-specs/blob/a70ffe7159e60a9548679d2924b24fc96da929a5/specification/recoveryservicesbackup/resource-manager/Microsoft.RecoveryServices/stable/2016-12-01/backupManagement.json

**Review by marstr — APPROVED** [review]

**Comment by marstr** [comment]
@sonathan are you ready for this to be merged?

**Comment by sonathan** [comment]
@marstr Thanks for approving! Yes please go ahead and merge this. Thanks a lot for reviewing this.

**Comment by lmazuel** [comment]
@AutorestCI rebuild azure-sdk-for-python

**Comment by lmazuel** [comment]
@AutorestCI rebuild azure-sdk-for-python

**Pull request merged**
